lib/client/src/client.rs

```diff
@@ -1057,58 +1057,107 @@ impl MangoClient {
-                            perp_market: perp.address,
-                            bids: perp.bids,
-                            asks: perp.asks,
-                            event_queue: perp.event_queue,
-                            oracle: perp.oracle,
-                        },
-                        None,
-                    );
-                    ams.extend(health_remaining_metas.into_iter());
-                    ams
-                },
-                data: anchor_lang::InstructionData::data(
-                    &mango_v4::instruction::PerpPlaceOrderV2 {
-                        side,
-                        price_lots,
-                        max_base_lots,
-                        max_quote_lots,
-                        client_order_id,
-                        order_type,
-                        reduce_only,
-                        expiry_timestamp,
-                        limit,
-                        self_trade_behavior,
+        let ix = Instruction {
+            program_id: mango_v4::id(),
+            accounts: {
+                let mut ams = anchor_lang::ToAccountMetas::to_account_metas(
+                    &mango_v4::accounts::PerpPlaceOrder {
+                        group: self.group(),
+                        account: self.mango_account_address,
+                        owner: self.owner(),
+                        perp_market: perp.address,
+                        bids: perp.bids,
+                        asks: perp.asks,
+                        event_queue: perp.event_queue,
+                        oracle: perp.oracle,
                     },
-                ),
+                    None,
+                );
+                ams.extend(health_remaining_metas.into_iter());
+                ams
             },
+            data: anchor_lang::InstructionData::data(&mango_v4::instruction::PerpPlaceOrderV2 {
+                side,
+                price_lots,
+                max_base_lots,
+                max_quote_lots,
+                client_order_id,
+                order_type,
+                reduce_only,
+                expiry_timestamp,
+                limit,
+                self_trade_behavior,
+            }),
+        };
+
+        ixs.push(
+            ix,
             self.instruction_cu(health_cu)
                 + self.context.compute_estimates.cu_per_perp_order_match * limit as u32,
         );
 
         Ok(ixs)
     }
 
+    async fn replace_perp_market_if_needed(
+        &self,
+        account: &MangoAccountValue,
+        perk_market_index: PerpMarketIndex,
+    ) -> anyhow::Result<Option<(PreparedInstructions, MangoAccountValue)>> {
+        let context = &self.context;
+        let settle_token_index = context.perp(perk_market_index).settle_token_index;
+
+        let mut account = account.clone();
+        let enforce_position_result =
+            account.ensure_perp_position(perk_market_index, settle_token_index);
+
+        if enforce_position_result.is_err() {
```

Comment:
would prefer if this checked for NoFreePerpPositionIndex, in case enforce_perp_position() changes

minor: I also typically prefer to return early over indenting the remaining function